coreax/inverses.py

```diff
@@ -0,0 +1,312 @@
+    if (power_iterations <= 0.0) or not isinstance(power_iterations, int):
+        raise ValueError("'power_iterations' must be a positive integer")
+
+    standard_gaussian_draws = jr.normal(
+        random_key, shape=(array.shape[0], oversampling_parameter)
+    )
+
+    # QR decomposition to find orthonormal array with range approximating range of array
+    approximate_range = array @ standard_gaussian_draws
+    q, _ = jnp.linalg.qr(approximate_range)
+
+    # Power iterations for improved accuracy
+    for _ in range(power_iterations):
+        approximate_range_ = array.T @ q
+        q_, _ = jnp.linalg.qr(approximate_range_)
+        approximate_range = array @ q_
+        q, _ = jnp.linalg.qr(approximate_range)
+
+    # Form the low rank array, compute its exact eigendecomposition and ortho-normalise
+    # the eigenvectors.
+    array_approximation = q.T @ array @ q
+    approximate_eigenvalues, eigenvectors = jnp.linalg.eigh(array_approximation)
+    approximate_eigenvectors = q @ eigenvectors
+
+    return approximate_eigenvalues, approximate_eigenvectors
+
+
+class RandomisedEigendecompositionApproximator(RegularisedInverseApproximator):
+    """
+    Approximate regularised inverse of a Hermitan array via random eigendecomposition.
+
+    When a dataset is very large, computing the regularised inverse of an array
+    can be very time-consuming. Instead, this property can be approximated by
+    various methods. :class:`RandomisedEigendecompositionApproximator` is a class that
+    does such an approximation using a randomised eigendecomposition.
+
+    Using (:cite:`halko2009randomness` Algorithm 4.4. and 5.3) we approximate the
+    eigendecomposition of a matrix. The parameters `oversampling_parameter`
+    and `power_iterations` present a trade-off between speed and approximation quality.
+    See :cite:`halko2009randomness`for discussion on choosing sensible parameters, the
+    defaults chosen here are cautious.
+
+    :param random_key: Key for random number generation
+    :param oversampling_parameter: Number of random columns to sample; the larger the
+        oversampling_parameter, the more accurate, but slower the method will be
+    :param power_iterations: Number of power iterations to do; the larger the
+        power_iterations, the more accurate, but slower the method will be
+    :param rcond: Cut-off ratio for small singular values of the `array`. For the
+        purposes of rank determination, singular values are treated as zero if they are
+        smaller than rcond times the largest singular value of a. The default value of
+        None will use the machine precision multiplied by the largest dimension of
+        the array. An alternate value of -1 will use machine precision.
+
+    """
+
+    oversampling_parameter: int = 25
+    power_iterations: int = 1
+    rcond: Optional[float] = None
+
+    def __check_init__(self):
+        """Validate rcond input."""
+        if self.rcond is not None:
```

Comment:
In this case we do need to check this ourselves because we are not calling any in-built functions that use and check values of rcond. We use it ourselves directly.